merge_sketch_with_bootloader.go

```diff
@@ -80,70 +83,85 @@ func (s *MergeSketchWithBootloader) Run(ctx *types.Context) error {
+func merge(builtSketchPath, bootloaderPath, mergedSketchPath string) error {
 
-	// skip until we find a line full of FFFFFF (except address and checksum)
-	for i, row := range bootloader {
-		if hexLineOnlyContainsFF(row) {
-			realBootloader = bootloader[i:len(bootloader)]
-			break
-		}
+	if filepath.Ext(bootloaderPath) == ".bin" {
+		bootloaderPath = strings.TrimSuffix(bootloaderPath, ".bin") + ".hex"
 	}
 
-	// drop all "empty" lines
-	for i, row := range realBootloader {
-		if !hexLineOnlyContainsFF(row) {
-			realBootloader = realBootloader[i:len(realBootloader)]
-			break
-		}
+	bootFile, err := os.Open(bootloaderPath)
+	if err != nil {
+		return err
 	}
+	defer bootFile.Close()
 
-	if len(realBootloader) == 0 {
-		// we didn't find any line full of FFFF, thus it's a standalone bootloader
-		realBootloader = bootloader
+	mem_boot := gohex.NewMemory()
+	err = mem_boot.ParseIntelHex(bootFile)
+	if err != nil {
+		return errors.New(bootFile.Name() + " " + err.Error())
 	}
 
-	return realBootloader
-}
-
-func merge(builtSketchPath, bootloaderPath, mergedSketchPath string) error {
-	sketch, err := utils.ReadFileToRows(builtSketchPath)
+	buildFile, err := os.Open(builtSketchPath)
 	if err != nil {
-		return i18n.WrapError(err)
+		return err
 	}
-	sketch = sketch[:len(sketch)-2]
+	defer buildFile.Close()
 
-	bootloader, err := utils.ReadFileToRows(bootloaderPath)
+	mem_sketch := gohex.NewMemory()
+	err = mem_sketch.ParseIntelHex(buildFile)
 	if err != nil {
-		return i18n.WrapError(err)
+		return errors.New(buildFile.Name() + " " + err.Error())
 	}
 
-	realBootloader := extractActualBootloader(bootloader)
+	mem_merge := gohex.NewMemory()
+	initial_address := uint32(math.MaxUint32)
+	last_address := uint32(0)
+
+	for _, segment := range mem_boot.GetDataSegments() {
+		err = mem_merge.AddBinary(segment.Address, segment.Data)
+		if err != nil {
+			continue
+		} else {
+			if segment.Address < initial_address {
+				initial_address = segment.Address
+			}
+			if segment.Address+uint32(len(segment.Data)) > last_address {
+				last_address = segment.Address + uint32(len(segment.Data))
+			}
+		}
+	}
+	for _, segment := range mem_sketch.GetDataSegments() {
+		err = mem_merge.AddBinary(segment.Address, segment.Data)
+		if err != nil {
+			continue
+		}
+		if segment.Address < initial_address {
+			initial_address = segment.Address
+		}
+		if segment.Address+uint32(len(segment.Data)) > last_address {
+			last_address = segment.Address + uint32(len(segment.Data))
+		}
+	}
 
-	for _, row := range realBootloader {
-		sketch = append(sketch, row)
+	mergeFile, err := os.Create(mergedSketchPath)
+	if err != nil {
+		return err
 	}
+	defer mergeFile.Close()
+
+	mem_merge.DumpIntelHex(mergeFile, 16)
+
+	mergedSketchPathBin := strings.TrimSuffix(mergedSketchPath, ".hex") + ".bin"
 
-	return utils.WriteFile(mergedSketchPath, strings.Join(sketch, "\n"))
+	bytes := mem_merge.ToBinary(initial_address, last_address-initial_address, 0xFF)
```

Comment:
This must only be executed if address space is continuous (otherwise we can get a huge file as result)